docs/tech_specs/dup_v2_alert_widget_backend.md

```diff
@@ -0,0 +1,141 @@
+
+[DUP alert spec](https://www.notion.so/mbta-downtown-crossing/DUP-Alert-Widget-Specification-a82acff850ed4f2eb98a04e5f3e0fe52)
+
+# Motivation
+
+[motivation]: #motivation
+
+Converting the DUP app from v1 to v2 requires a different approach to alerts and how they are displayed on the screen.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The DUP alerts widget needs to resolve a scenario specific to DUP screens: the screen "rotates" between three separate pages. When an applicable[^1] alert exists, each rotation has specific logic that determines which "type" of alert widget to display: `partial`[^2] or `takeover`[^3].
+
+### CandidateGenerator
+
+`CandidateGenerator.Dup` is responsible for deciding what alert should appear on a screen. It fetches all active alerts affecting the screen's `stop_ids` listed in each section of the `primary_departures` config, filters out all alerts that do not directly affect the current stop (no downstream alerts), and selects the alert with the highest priority[^4]. If an alert (or alerts if configured with two sections) meets all criteria, the `CandidateGenerator` will then create three `WidgetInstance`s for the alert(s) (one for each rotation).
+
+### WidgetInstance
+
+The DUP alert `WidgetInstance` is responsible for choosing alert type and serializing the alert data for the frontend to consume. The alert type will chosen based on rotation and how impactful the alert is.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### CandidateGenerator
+
+`CandidateGenerator.Dup` is responsible for fetching relevant alerts from the API. The API query will contain an `opts` array for each section:
+
+```
+opts = [
+      stop_ids: <stop_ids from config>,
+      route_ids: <route_ids from config>,
+      route_types: ~w[light_rail subway]a
+    ]
+```
+
+Each section will make a separate alerts fetch. With the two sets of results from these queries, we will apply a filter to eliminate alerts we do not want on screens: `Enum.filter(&relevant?/1)`. `relevant?/1` will return a list of active and applicable alerts[^1]. After we have a list (or lists if there are two sections) of alerts that are eligible for display, we choose the alert to display for each section based on a priority[^4]. The pipeline for these steps for each section will be the following:
+
+```
+first_section_alerts =
+    opts
+    |> Alert.fetch()
+    |> Enum.filter(&relevant?/1)
+    |> choose_alert()
+
+second_section_alerts = []
+if length(sections) == 2 do
+    second_section_alerts =
+        opts
+        |> Alert.fetch()
+        |> Enum.filter(&relevant?/1)
+        |> choose_alert()
+end
+
+alerts = first_section_alerts ++ second_section_alerts
+```
+
+If screen is configured with only one section, `alerts` will contain up to one item. If there are two sections configured, `alerts` will contain up to two items.
```

Comment:
Why the different behavior for screens with one vs two sections? In general, we'll still only show one alert _per screen_, even if there are multiple sections.

One known exception: We do have some special cases where we handle combinations of multiple shuttle/suspension alerts extending west starting at Kenmore, which would require being able to see all of those alerts at once.

Reply:
That behavior is how v1 currently deals with an alert being active in each section [here](https://github.com/mbta/screens/blob/master/lib/screens/dup_screen_data/data.ex#L87). If there are two sections in primary departures and each one has a relevant alert, a fullscreen alert takes the place of all 3 rotations (as of this comment, you can see that in place [here](https://screens-dev-green.mbtace.com/screen/DUP-Haymarket)). Outside of Kenmore (which we know is its own can of worms), this will only be in play for Haymarket. It is the only other screen that is configured with 2 subway sections, which is the only mode we fetch alerts for. 

In my opinion, this does feel a little aggressive. For example, if there are two delays up to 30 minutes at Haymarket, one for OL one for GL), all 3 rotations show a fullscreen alert and no departures are displayed. So I would understand wanting to move away from this logic/tweaking it so this only happens when all predictions are suspended.

Reply:
It sounds like this is a divergence from v1, then, but (at least for now) in v2 we'll just show one alert per screen.

I can imagine deciding later that we might want to add some more cases -- for example we might check whether the current alerts eliminate all service at a station and do something different -- but none of that is decided yet, and that would come after we ship a first pass at the widgetized DUP app.

Reply:
When you say this is a divergence, do you mean the logic in the doc or the suggestion in my comment? In v2, I mentioned that we still will only mention one alert even if two are present in the `alerts` attribute. The second alert is only used to determine if we need a fullscreen alert on all rotations.

Reply:
I think this result is actually a bug. When I pull up Haymarket with these alerts, I'm getting an Elixir error. We do not have a function set up to handle a `:boundary` alert on a multi-line screen. That scenario was only set up for `:inside`. Based on that, I'm guessing we do _not_ want this. 

The scenario I mentioned above where there is an alert in each section on a screen, we see the following:

![Screenshot 2023-02-02 at 12 57 01 PM](https://user-images.githubusercontent.com/10713153/216404374-9fed52a9-c754-4a9e-9d43-6052f4778c2a.png)

I think in most cases, this could be ok. Mostly if predictions are suspended because of the alerts. However, this image was captured with two delay alerts. While the text is true, I'm not sure a takeover on all rotations is the appropriate response to that scenario. Regardless, this scenario is the reason for the paragraph that was originally commented on in this thread.